aten/src/ATen/native/TensorShape.cpp

```diff
@@ -436,10 +443,45 @@ Tensor & transpose_(Tensor & self, int64_t dim0, int64_t dim1) {
   return self.as_strided_(sizes, strides);
 }
 
-Tensor & t_(Tensor & self) {
-  if (self.ndimension() != 2) {
-    AT_ERROR("t_() expects a 2D tensor, but self is ", self.ndimension());
+Tensor transpose(const Tensor & self, int64_t dim0, int64_t dim1) {
+  auto ndims = self.dim();
+  dim0 = maybe_wrap_dim(dim0, ndims);
+  dim1 = maybe_wrap_dim(dim1, ndims);
+  if (dim0 == dim1) {
+    return self;
   }
+
+  if (self.is_sparse()) {
+    Tensor self_clone = self.clone();  // yes, this is what THS does
+    return sparse_transpose_(self_clone, dim0, dim1);
+  }
+
+  std::vector<int64_t> strides(self.strides());
+  std::vector<int64_t> sizes(self.sizes());
+  std::swap(strides[dim0], strides[dim1]);
+  std::swap(sizes[dim0], sizes[dim1]);
+  return self.as_strided(sizes, strides);
+}
+
+static void check_t(const Tensor& self, const char *fn) {
+  if (self.is_sparse()) {
+    int64_t nDimI = self._dimI();
+    int64_t nDimV = self._dimV();
+    if (!(nDimI == 2 && nDimV == 0)) {
+      AT_ERROR(fn, " expects a 2D sparse tensor, but self is ", nDimI, "D indices and ", nDimV, "D values");
+    }
+  } else if (self.ndimension() != 2) {
+    AT_ERROR(fn, " expects a 2D tensor, but self is ", self.ndimension());
```

Comment:
nit: `AT_ERROR(fn, " expects a 2D tensor, but self is ", self.ndimension(), "D");`